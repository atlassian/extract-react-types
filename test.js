// @flow
"use strict";

const extractReactTypes = require("./");
const stripIndent = require("strip-indent");

const TESTS = [
  {
    name: "flow boolean",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: boolean }> {
      // ...
    }
  `
  },
  {
    name: "flow string",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: string }> {
      // ...
    }
  `
  },
  {
    name: "flow number",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: number }> {
      // ...
    }
  `
  },
  {
    name: "flow function",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: (string, number) => string }> {
      // ...
    }
  `
  },
  {
    name: "flow function named params",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: (name: string, age: number) => void }> {
      // ...
    }
  `
  },
  {
    name: "flow function unnamed params",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: (string, number) => void }> {
      // ...
    }
  `
  },
  {
    name: "flow function unnamed params w/ object",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: ({ prop: string }) => void }> {
      // ...
    }
  `
  },
  {
    name: "flow function return type",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: (number) => void }> {
      // ...
    }
  `
  },
  {
    name: "flow function type",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: Function }> {
      // ...
    }
  `
  },
  {
    name: "flow maybe",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: ?string }> {
      // ...
    }
  `
  },
  {
    name: "flow array",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: Array<boolean> }> {
      // ...
    }
  `
  },
  {
    name: "flow union",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: boolean | number }> {
      // ...
    }
  `
  },
  {
    name: "flow union literals",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ age: '25' | 30 }> {
      // ...
    }
  `
  },
  {
    name: "flow mixed",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ age: mixed }> {
      // ...
    }
  `
  },
  {
    name: "flow any",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ age: any }> {
      // ...
    }
  `
  },
  {
    name: "flow generic class",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<A> {
      name: A;
    }
  `
  },
  {
    name: "flow type",
    typeSystem: "flow",
    code: `
    type ComponentProps = {
      foo: number,
      bar: boolean
    };
    class Component extends React.Component<ComponentProps> {

    }
  `
  },
  {
    name: "flow type identifier",
    typeSystem: "flow",
    code: `
    import * as React from 'react';

    type Props = {
      children: React.Node,
    };

    class Component extends React.Component<Props> {

    }
  `
  },
  {
    name: "flow exported type alias reference",
    typeSystem: "flow",
    code: `
    export type Foo = boolean;
    class Component extends React.Component<{ foo: Foo }> {

    }
  `
  },
  {
    name: "flow type",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{foo: number & string}> {

    }
  `
  },
  {
    name: "intersection type",
    typeSystem: "flow",
    code: `
  type BaseProps = { bar: string }

  class Component extends React.Component<BaseProps & {
    isDefaultChecked: boolean,
  }> {
  }
  `
  },
  {
    name: "with spread in type annotation",
    typeSystem: "flow",
    code: `
  type BaseProps = { foo: string }
  type Props = {
    ...BaseProps,
    isDefaultChecked: boolean,
  }
  class Component extends React.Component<Props> {
  }
  `
  },
  {
    name: "flow array union",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: Array<boolean | number> }> {
      // ...
    }
  `
  },
  {
    name: "flow void",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo: void }> {
      // ...
    }
  `
  },
  {
    name: "flow optional",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo?: number }> {
    }
  `
  },
  {
    name: "flow no React component",
    typeSystem: "flow",
    code: `
    class FooComponent extends Bar<{ foo?: number }> {
    }
  `
  },
  {
    name: "flow different React component",
    typeSystem: "flow",
    code: `
    import {Component} from 'react';

    class FooComponent extends Component<{ foo?: number }> {
    }

    class BarComponent extends React.Component<{ foo?: number }> {
    }
  `
  },
  {
    name: "flow import { type }",
    typeSystem: "flow",
    code: `
    import React, { type Node } from 'react';

    type Props = {
      children: Node,
    };

    class Component extends React.Component<Props> {

    }
  `
  },
  {
    name: "ts boolean",
    typeSystem: "typescript",
    code: `
    class Component extends React.Component<{ foo: boolean }> {

    }
  `
  },
  {
    name: "ts string",
    typeSystem: "typescript",
    code: `
    class Component extends React.Component<{ foo: string }> {

    }
  `
  },
  {
    name: "ts object",
    typeSystem: "typescript",
    code: `
    class Component extends React.Component<{foo: { name: string }, bar: number, verified: boolean}> {

    }
  `
  },
  {
    name: "ts interface",
    typeSystem: "typescript",
    code: `
    interface ComponentProps {
      foo: string;
      bar: number;
    }

    class Component extends React.Component<ComponentProps> {
      // ...
    }
  `
  },
  {
    name: "ts function",
    typeSystem: "typescript",
    code: `
    class Component extends React.Component<{foo: (string, boolean) => number}> {
      // ...
    }
  `
  },
  {
    name: "ts array",
    typeSystem: "typescript",
    code: `
    class Component extends React.Component<{foo: Array<number>}> {
      // ...
    }
  `
  },
  {
    name: "ts union",
    typeSystem: "typescript",
    code: `
    class Component extends React.Component<{foo: number | string}> {
      // ...
    }
  `
  },
  {
    name: "ts any",
    typeSystem: "typescript",
    code: `
    class Component extends React.Component<{foo: any}> {
      // ...
    }
  `
  },
  {
    name: "ts literals",
    typeSystem: "typescript",
    code: `
    class Component extends React.Component<{foo: '25' | 30}> {

    }
  `
  },
  {
    name: "ts optional",
    typeSystem: "typescript",
    code: `
    class Component extends React.Component<{foo?: string}> {

    }
  `
  },
  {
    name: "ts void",
    typeSystem: "typescript",
    code: `
    class Component extends React.Component<{foo: void}> {

    }
  `
  },
  {
    name: "ts no react component",
    typeSystem: "typescript",
    code: `
    class Component extends Foo<{foo: void}> {

    }
  `
  },
  {
    name: "ts different React components",
    typeSystem: "typescript",
    code: `
    import {Component} from 'react';

    class FooComponent extends Component<{foo: void}> {

    }

    class BarComponent extends React.Component<{foo: void}> {

    }
  `
  },
  {
    name: "ts tuple",
    typeSystem: "typescript",
    code: `
    class Component extends React.Component<{foo: [string, number]}> {

    }
  `
  },
  {
    name: "ts enum",
    typeSystem: "typescript",
    code: `
    enum Color {Red, Green, Blue};
    class Component extends React.Component<{foo: Color}> {

    }
  `
  },
  {
    name: "test",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ foo?: Object }> {}
  `
  },
  {
    name: "test get defaultProps",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ a: string }> {
      defaultProps = {
        a: 'a'
      }
    }
  `
  },
  {
    name: "test defaultProp as JSXElement",
    typeSystem: "flow",
    code: `
      import type { Node } from 'react';
      const Icon = <div></div>
      class Component extends React.Component<{a: Node }> {
        defaultProps = {
          a: <Icon name={"test icon"}/>
        }
      }
    `
  },
  {
    name: "test defaultProp as JSXElement with member expression",
    typeSystem: "flow",
    code: `
    import type { Node } from 'react';
    const Icon = <div></div>
    const componentObj = { Icon }
    class Component extends React.Component<{a: Node }> {
      defaultProps = {
        a: <componentObj.Icon name={"test icon"}/>
      }
    }
    `
  },
  {
    name: "test defaultProp as JSXElement with multiple props",
    typeSystem: "flow",
    code: `
      import type { Node } from 'react';
      const Icon = <div></div>
      class Component extends React.Component<{a: Node}>{
        defaultProps = {
          a: <Icon name="test icon" iconType="avatar" />
        }
      }
    `
  }, {
      name: "test defaultProp as JSXElement with JSXExpressionContainer",
      typeSystem: "flow",
      code: `
        import type { Node } from 'react';
        const Icon = <div></div>
        class Component extends React.Component<{a: Node}>{
          defaultProps = {
            a: <Icon name={"test icon"} />
          }
        }
      `
    },
  {
    name: "spread element ",
    typeSystem: "flow",
    code: `

    const something = { a: true }
    class Component extends React.Component<{ a: boolean, b: boolean }> {
      defaultProps = {
        ...something,
        b: false
      }
    }
  `
  },
  {
    name: "string literal",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ a: string }> {
      defaultProps = {
        a: 'stringVal'
      }
    }
  `
  },
  {
    name: "numeric literal",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ a: number }> {
      defaultProps = {
        a: 5
      }
    }
  `
  },
  {
    name: "null literal",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ a?: number }> {
      defaultProps = {
        a: null
      }
    }
  `
  },
  {
    name: "boolean literal",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ a: boolean }> {
      defaultProps = {
        a: true
      }
    }
  `
  },
  {
    name: "array expression",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ a: Array<string> }> {
      defaultProps = {
        a: ['a']
      }
    }
  `
  },
  {
    name: "binary expression",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ a: number }> {
      defaultProps = {
        a: 3 + 5
      }
    }
  `
  },
  {
    name: "member expression",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ a: () => mixed }> {
      defaultProps = {
        a: b.c
      }
    }
  `
  },
  {
    name: "defined member expression",
    typeSystem: "flow",
    code: `
    const b = {
      c: (a: string, b: string): number => {}
    }
    class Component extends React.Component<{ a: () => mixed }> {
      defaultProps = {
        a: b.c
      }
    }
  `
  },
  {
    name: "imported member expression default",
    typeSystem: "flow",
    code: `
    import c from 'somewhere-awesome'
    class Component extends React.Component<{ a: () => mixed }> {
      defaultProps = {
        a: b.c
      }
    }
  `
  },
  {
    name: "imported member expression named",
    typeSystem: "flow",
    code: `
    import { c } from 'somewhere-awesome'
    class Component extends React.Component<{ a: () => mixed }> {
      defaultProps = {
        a: b.c
      }
    }
  `
  },
  {
    name: "imported member expression named alias",
    typeSystem: "flow",
    code: `
    import { d as c } from 'somewhere-awesome'
    class Component extends React.Component<{ a: () => mixed }> {
      defaultProps = {
        a: b.c
      }
    }
  `
  },
  {
    name: "arrow function expression",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ a: () => mixed }> {
      defaultProps = {
        a: (): number => 5
      }
    }
  `
  },
  {
    name: "function expression",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ a: () => mixed }> {
      defaultProps = {
        a: function (): number {}
      }
    }
  `
  },
  {
    name: "exists type annotation",
    typeSystem: "flow",
    code: `
    class Component extends React.Component<{ a: () => mixed }> {
      defaultProps = {
        a: function (): * {}
      }
    }
  `
  }
];

for (let testCase of TESTS) {
  let testFn;

  if (testCase.only) {
    testFn = test.only;
  } else if (testCase.skip) {
    testFn = test.skip;
  } else {
    testFn = test;
  }

  testFn(testCase.name, () => {
    let code = stripIndent(testCase.code);
    let result = extractReactTypes(code, testCase.typeSystem);
    expect(result).toMatchSnapshot();
  });
}
