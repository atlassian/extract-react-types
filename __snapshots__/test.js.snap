// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comments 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "description": "* Comment
  * asdf
      asdf
  
Comment2",
          "key": "foo",
          "kind": "property",
          "leadingComments": Array [
            Object {
              "end": 54,
              "loc": SourceLocation {
                "end": Position {
                  "column": 4,
                  "line": 6,
                },
                "start": Position {
                  "column": 2,
                  "line": 3,
                },
              },
              "start": 18,
              "type": "CommentBlock",
              "value": "* Comment
  * asdf
      asdf
  ",
            },
            Object {
              "end": 68,
              "loc": SourceLocation {
                "end": Position {
                  "column": 13,
                  "line": 7,
                },
                "start": Position {
                  "column": 2,
                  "line": 7,
                },
              },
              "start": 57,
              "type": "CommentLine",
              "value": " Comment2",
            },
          ],
          "optional": false,
          "value": Object {
            "kind": "boolean",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow any 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "age",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "any",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow array 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": undefined,
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow array union 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": undefined,
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow boolean 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "boolean",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow different React component 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": true,
          "value": Object {
            "kind": "number",
          },
        },
      ],
    },
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": true,
          "value": Object {
            "kind": "number",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow exported type alias reference 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "boolean",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow function 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "function",
            "parameters": Array [
              Object {
                "kind": "string",
              },
              Object {
                "kind": "number",
              },
            ],
            "returnType": Object {
              "kind": "string",
            },
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow function named params 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "function",
            "parameters": Array [
              Object {
                "kind": "string",
              },
              Object {
                "kind": "number",
              },
            ],
            "returnType": Object {
              "kind": "void",
            },
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow function return type 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "function",
            "parameters": Array [
              Object {
                "kind": "number",
              },
            ],
            "returnType": Object {
              "kind": "void",
            },
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow function type 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": undefined,
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow function unnamed params 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "function",
            "parameters": Array [
              Object {
                "kind": "string",
              },
              Object {
                "kind": "number",
              },
            ],
            "returnType": Object {
              "kind": "void",
            },
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow function unnamed params w/ object 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "function",
            "parameters": Array [
              Object {
                "kind": "object",
                "props": Array [
                  Object {
                    "key": "prop",
                    "kind": "property",
                    "optional": false,
                    "value": Object {
                      "kind": "string",
                    },
                  },
                ],
              },
            ],
            "returnType": Object {
              "kind": "void",
            },
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow generic class 1`] = `
Object {
  "classes": Array [
    undefined,
  ],
  "kind": "program",
}
`;

exports[`flow import { type } 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "children",
          "kind": "property",
          "optional": false,
          "value": Object {
            "importKind": "type",
            "kind": "import",
            "moduleSpecifier": "react",
            "name": "Node",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow maybe 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "arguments": Object {
              "kind": "string",
            },
            "kind": "nullable",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow mixed 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "age",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "mixed",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow no React component 1`] = `
Object {
  "classes": Array [],
  "kind": "program",
}
`;

exports[`flow number 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "number",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow optional 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": true,
          "value": Object {
            "kind": "number",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow string 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "string",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow type 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "number",
          },
        },
        Object {
          "key": "bar",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "boolean",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow type 2`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "intersection",
            "types": Array [
              Object {
                "kind": "number",
              },
              Object {
                "kind": "string",
              },
            ],
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow type identifier 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "children",
          "kind": "property",
          "optional": false,
          "value": undefined,
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow union 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "union",
            "types": Array [
              Object {
                "kind": "boolean",
              },
              Object {
                "kind": "number",
              },
            ],
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow union literals 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "age",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "union",
            "types": Array [
              Object {
                "kind": "stringLiteral",
                "value": "25",
              },
              Object {
                "kind": "numberLiteral",
              },
            ],
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`flow void 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "key": "foo",
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "void",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts any 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "any",
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts array 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        undefined,
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts boolean 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "boolean",
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts different React components 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "void",
        },
      ],
    },
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "void",
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts enum 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        undefined,
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts function 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "function",
          "parameters": Array [
            undefined,
            undefined,
          ],
          "returnType": Object {
            "kind": "number",
          },
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts interface 1`] = `
Object {
  "classes": Array [
    undefined,
  ],
  "kind": "program",
}
`;

exports[`ts literals 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "union",
          "types": Array [
            Object {
              "kind": "literal",
              "value": "25",
            },
            Object {
              "kind": "literal",
              "value": 30,
            },
          ],
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts no react component 1`] = `
Object {
  "classes": Array [],
  "kind": "program",
}
`;

exports[`ts object 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "object",
          "props": Array [
            Object {
              "kind": "string",
            },
          ],
        },
        Object {
          "kind": "number",
        },
        Object {
          "kind": "boolean",
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts optional 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "string",
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts string 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "string",
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts tuple 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "tuple",
          "types": Array [
            Object {
              "kind": "string",
            },
            Object {
              "kind": "number",
            },
          ],
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts union 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "union",
          "types": Array [
            Object {
              "kind": "number",
            },
            Object {
              "kind": "string",
            },
          ],
        },
      ],
    },
  ],
  "kind": "program",
}
`;

exports[`ts void 1`] = `
Object {
  "classes": Array [
    Object {
      "kind": "object",
      "props": Array [
        Object {
          "kind": "void",
        },
      ],
    },
  ],
  "kind": "program",
}
`;
